### PR TITLE
feat(rect.rs): implement Rows and Columns iterators in Rect

### PR DESCRIPTION
**Pull request by BogdanPaul15**
Signed-off-by: Prisacaru Bogdan-Paul <paulbp68@gmail.com>

#752

This PR will allow users to use iterators for Rows and Columns in Rects in order to manage rows and columns. During iteration, each structure produces `Rect` instances representing each row and column within the given Rect.

### CODE REVIEW

**Review by Valentin271 — COMMENTED** [review]
I'm not sure this is what joshka meant when creating the issue.

I think you wanted to do the following right?
```diff
- for x in area.left()..area.right() { 
+ for x in area.columns() { 
-   for y in area.top()..area.bottom() { 
+   for y in area.rows() { 
    buf.get_mut(x, y).reset(); 
  } 
} 
```

The way I understood the issue, the iterators were supposed to return `Rect`s representing the rows and column. So we could work on that and, for example, get the columns of a row (so a `Rect`).

```rust
let area = Rect(0, 0, 10, 5);
for row in area.rows() {
  // row is Rect { x: 0, y: 0, width: 10, height: 1}, note height of one because it's a row
  // on second iteration this will become Rect { x: 0, y: 1, width: 10, height: 1}
  // up to Rect { x: 0, y: 9, width: 10, height: 1}, y of 9 because there is 10 rows starting from 0

  // and then here because row is a `Rect` we can call `columns` again
  for cell in row.columns() {
    // which will give Rect { x: 0, y: 0, width: 1, height: 1}, width is one because it's a columns, height is one because we iterate over a row
    // then Rect { x: 1, y: 0, width: 1, height: 1}, incrementing x up to 9
    // and at the end Rect { x: 9, y: 9, width: 1, height: 1}, the last cell on the bottom right

    // here we can call rect.x and rect.y because the resulting rect, when iterating over rows and columns, is 1 by 1, so a cell
    buf.get_mut(cell.x, cell.y).reset();
  }
}
```

That was my understanding, let's wait for @joshka's input

**Comment by BogdanPaul15** [comment]
I understand your point. I will wait for @joshka. Thank you!

**Review by joshka — CHANGES_REQUESTED** [review]
@Valentin271 was right. I'd expect each element of the iterator to be a single height or single width Rect.

For the tests, simplify them to compare rows().collect() to a hard coded vec of rects. This will visually show that the x or y index is changing, but the width or height is always 1.

**Comment by BogdanPaul15** [comment]
From my understanding, this is what you needed. Please let me know if I am wrong.

File: src/layout/rect.rs
Comment (Valentin271) [inline src/layout/rect.rs:37]: You don't need the current column in the row iterator.

File: src/layout/rect.rs
Comment (Valentin271) [inline src/layout/rect.rs:63]: Same here you don't need the current_row in the columns iterator.

File: src/layout/rect.rs
Comment (Valentin271) [inline src/layout/rect.rs:47]: You can use `Rect::bottom()` here to make this easier to maintain and avoid panics.
```suggestion
        if self.current_row >= self.rect.bottom() {
```

File: src/layout/rect.rs
Comment (Valentin271) [inline src/layout/rect.rs:75]: Similarly, you can use `Rect::right()` here.
```suggestion
        if self.current_column >= self.rect.right() {
```

**Review by Valentin271 — CHANGES_REQUESTED** [review]
Other than my few comments looks good.

**Comment by Valentin271** [comment]
It looks like you tried to sign your commit (as shows the PR description), but the github interface doesn't show it, could check that?

You have some information here https://github.com/ratatui-org/ratatui/blob/main/CONTRIBUTING.md#sign-your-commits.

If you need help feel free to ask.

**Comment by BogdanPaul15** [comment]
@Valentin271 thank you for your feedback. Should I commit these changes? Also, yes, I've tried to sign my commits, my machine is already configured with gpg keys, but I don't know why it doesn't work.

**Comment by Valentin271** [comment]
Have you looked at https://commitizen-tools.github.io/commitizen/getting_started/#sign-off-the-commit? (weird though I don't remember having to do that when I used it)

You can either commit the proposed changes from the interface (will be committed as me though) or commit them yourself as usual, as you wish.

**Review by Valentin271 — APPROVED** [review]
Looks all good, code and signing. Thanks!

I'll wait for a second approval on this.

**Review by joshka — APPROVED** [review]
Looks great.

**Comment by joshka** [comment]
To get this to merge, all the commits need to be signed, not just the last one. Can you please squash the commits locally into a single commit and force push to your branch?

image

Thanks for the PR :)

**Comment by BogdanPaul15** [comment]
Ok, I'm trying to squash into a single commit and force push.

**Comment by BogdanPaul15** [comment]
I need some help. When I try to squash my commits, I am getting others commits than mine. Do I have to squash them too?

**Comment by Valentin271** [comment]
I'd suggest

```sh
// this will go back to your latest relevant commit
git reset --hard d9195580cc9eb2c941f66e34bb4630eee33d14c3

git rebase -i HEAD~4
// replace pick with squash for the last 3 commits, see attached screenshot 
// then the commit message will show up, edit as needed

// just to save the changes
git push -f --no-verify
```

Then sign your commit and force push.
If you can't/don't know how to sign an already existing commit, you can

```sh
// deletes the commit and put the changes back to staged
git reset --soft HEAD^
// then commit+sign as usual
// and force push
git push -f --no-verify
```

Might not be the most straightforward way but at least it should work.

---

`rebase -i` interface
image

**Comment by BogdanPaul15** [comment]
Thank you. I really don't lnow why this failed.

**Comment by Valentin271** [comment]
You're welcome, and thanks for your contribution!

**Pull request merged**
